### ISSUE TITLE
Parentheses can be removed (IDE0047) false-positives

### ISSUE THREAD

**Issue by shadow-cs** [dotnet/roslyn#29581]
Following code produces false-positive when `dotnet_style_parentheses_in_relational_binary_operators` is set to `always_for_clarity`.

```csharp
public static bool Foo(int a, int b, bool c)
{
  return (a == b) == c;
}
```

Since the rule is set to always for clarity, my assumption is that the code is indeed fine as-is.

I understand that the analyzer cannot always guess the parenthesis right in these situations and should ignore the statement rather than producing the message.

**Version Used**: 

* Visual Studio: 15.8.0
* Roslyn: 2.9.0.63208 (958f2354)

**Steps to Reproduce**:

1. Set `dotnet_style_parentheses_in_relational_binary_operators` to `always_for_clarity` in `.editorconfig`
2. Copy the code above to the editor

**Expected Behavior**:

No message, actually I'd prefer if the analyzer reported an issue if the parenthesis is missing (but this might be too strict to say for every scenario).

**Actual Behavior**:

IDE0047 is reported.

**Notes:**

Originally posted to: MicrosoftDocs/visualstudio-docs#1392

Related to: #29454

**Comment by CyrusNajmabadi** [comment]
> Since the rule is set to always for clarity, my assumption is that the code is indeed fine as-is.

This is not how the feature works.  "always for clarity" means: keep parentheses around if they help clarify precedence between different operators of that class.  i.e.  someone may not know that `*` and `+` have different precedence.  So if you have an operation containing both those ops, and you have "always for clarity" on, then it will require the parens to clarify thigns.

That does *not* apply when all the operators are the same.  For example, an expression containing only only `+`'s.  In that case, there is no precedence difference, so there is no need for parentheses to help with clarity.  indeed, that would be pretty annoying.  Because then, if you had that option on, the expression `a + b+ c + d` would become `((a + b) + c) + d`.  

That's what's happening here.  Your expression only uses `==`.  There is no clarity problem around the precedence.  So the parens are deemed as not being necessary for clarity.

**Comment by sharwell** [comment]
This was one of the edge cases found during the initial design review, but the pattern is exceedingly rare so it was deemed not problematic enough to change the design. The other case where it was called out is the following, which appeared a few more times:

```
(a == b) != (c == d)
```

**Comment by sharwell** [comment]
Tentatively marking this one as **Won't Fix**, but I'll leave it open until we can discuss it at a team design review.

**Comment by shadow-cs** [comment]
Thanks @sharwell for the clarification, I do understand this is a rarely used corner case (and suppressing the rule for such instances isn't a big problem).

**Comment by CyrusNajmabadi** [comment]
We originally had "ignore" as a potential choice as well for some of these options.  i.e. you could say "i don't want you to suggest adding *or* removing parens from relational expression".  We decided it was too minor a case to support, and you could get similar behavior by saying "Silent" as the severity of the option.

**Comment by CyrusNajmabadi** [comment]
> I think it does clarify == operator given some scenario (if I'm comparing three values of the same type for example, I wouldn't add the parenthesis, but when comparing multiple types and want to emphasize something, parenthesis is the obvious way to do it)

Could you expand on that?  Why do parentheses make it clear what's going on when types are different, but aren't necessary when types are the same?  I could actually see teh opposite argument.  When types are different (like in your case of bool and ints) it's clear the ints needs to be checked first, to create a bool, that will then be checked against another bool.  However, if they were all bools (the only real case where you could do multiple `==` and have things be the same type), then someone might not know the order of operations. 

That said, my personal take is: you don't really need parens for precedence clarity beween the same operator.  Because it's the same operator, there's no precedence question.

**Comment by shadow-cs** [comment]
> Could you expand on that? Why do parentheses make it clear what's going on when types are different, but aren't necessary when types are the same?

It is not about operator precedence in this case, but about emphasizing the intent of the comparison to make it clear which operations are bundled together. This is possible not why the feature was added in the first place. But as I said  before, I think of it as a way to make the code more readable and the analyzer should not get in the way of improving readability (we might not agree on the rules as this is highly tailored to each ones' personality ;-) ). That being said, if the analyzer was more configurable so we can achieve both outcomes, that would be awesome.

Like you @CyrusNajmabadi said if there were an option to ignore the relational expression, this might be a way, I don't want to ignore the entire rule as it spans too many operators and I want to check the other ones. Once again I do agree this is a corner case.

**Comment by CyrusNajmabadi** [comment]
> Like you @CyrusNajmabadi said if there were an option to ignore the relational expression, this might be a way

do you mean, having just an option to control the behavior for `==` and `!=` independent of the other relational operators?

**Comment by CyrusNajmabadi** [comment]
> Once again I do agree this is a corner case.

Agreed.  You can effectively only rnu into this by comparing one value to another, then comparing that result to a boolean.  It seems effectively too much of a corner case to warrant attention (esp. as i'm not even sure if there is a precedence/clarity issue here in the first place).  Why not just suppress the analyzer in this one location?  Then you can get the benefit for the rest of the code, without the 1-2 cases that aren't desirable by you? :)

**Comment by shadow-cs** [comment]
> do you mean, having just an option to control the behavior for == and != independent of the other relational operators?
This would be great.

> Why not just suppress the analyzer in this one location?
Already did ;-)

**Comment by CyrusNajmabadi** [comment]
Given how narrow the situation is, and given I think the behavior is acceptable.  I'm inclined to agree with: https://github.com/dotnet/roslyn/issues/29581#issuecomment-416905253

Thanks!

**Comment by sharwell** [comment]
Closing as won't fix based on the discussion above. It was a bit of a tough call in the original implementation, but the design we chose seems to do the best job of making it easy to explain an otherwise very complicated situation to users. We'll continue to watch for cases where the result of the analysis isn't as clear as users would want as well as the frequency of those cases, and if a specific pattern starts to appear that makes sense for edge case handling for all users then we could consider implementing it.

**Comment by jshunia** [comment]
I am requesting that this item be reopened.  With the 'always_for_clarity' option, the following code triggers the suggestion:

`int a = (x * y) / z;`

I understand that the parenthesis above do not change the order of operations, as the '*' and '/' operators have the same precedence, however, I prefer them for clarity when jumping between operators.  I suggest that when IDE0047 is set to 'always_for_clarity', then it should either never prompt, or it should only prompt for parenthesis that are truly unnecessary, e.g.:

`int a = ((x * y));`

Since the name is 'always_for_clarity', I think never prompting makes the most sense, but I could see arguments for prompting on the above.

**Comment by CyrusNajmabadi** [comment]
> I understand that the parenthesis above do not change the order of operations, as the '*' and '/' operators have the same precedence, however, I prefer them for clarity when jumping between operators.

You can get this behavior by setting this value:

![image](https://user-images.githubusercontent.com/4564579/46098274-dfd03280-c178-11e8-97bc-fb74cdc39f3a.png)

As you can see, with that setting, you get no suggestion prompts here:

![image](https://user-images.githubusercontent.com/4564579/46098364-13ab5800-c179-11e8-8054-a8d63466dc3b.png)

You are offered to still fix it if you bring up the lightbulb manually.  We still want to make it easy to remove if that's what you want.

Cheers!

**Comment by jshunia** [comment]
@CyrusNajmabadi:
Sorry for the delayed response, but I think maybe there is a bug here.  I have all of the parenthesis options set to 'Always for clarity', however I still see the suggestion with your code snippet.  

FYI I am using Visual Studio Professional 2017 Version 15.8.3.  I can provide additional details if required.

**Comment by CyrusNajmabadi** [comment]
Do you potentially have a .editorconfig file that is changing these values?  Can you show a screenshot of what you mean?

**Comment by mukunku** [comment]
I want to say that this is still an issue for me. If I use[ dotnet-format ](https://github.com/dotnet/format) for the below arithmetic operation it still removed the paranthesis:

![image](https://user-images.githubusercontent.com/4502154/162067217-010b482b-077a-47b4-8722-daccd0f0a25c.png)

```
  {
    "DocumentId": {
      "ProjectId": {
        "Id": "37496da6-5b94-44a4-8b57-12346ee12d0e"
      },
      "Id": "959567c8-4a89-4008-1234-cf5d286453d9"
    },
    "FileName": "User.cs",
    "FilePath": "User.cs",
    "FileChanges": [
      {
        "LineNumber": 79,
        "CharNumber": 30,
        "DiagnosticId": "IDE0047",
        "FormatDescription": "error IDE0047: Parentheses can be removed"
      }
    ]
  }
```

I have my editor config as follows:
```
dotnet_style_parentheses_in_arithmetic_binary_operators = always_for_clarity
dotnet_style_parentheses_in_other_binary_operators = always_for_clarity
dotnet_style_parentheses_in_relational_binary_operators = always_for_clarity
dotnet_style_parentheses_in_other_operators = never_if_unnecessary
```

**Comment by CyrusNajmabadi** [comment]
> I want to say that this is still an issue for me. If I use[ dotnet-format ](https://github.com/dotnet/format)for the below arithmetic operation it still removes the paranthesis:

The operators there have the same precedence (multiplicative), so they are not needed for clarity.   You can see that in this list: https://github.com/dotnet/csharpstandard/blob/draft-v6/standard/expressions.md#1142-operator-precedence-and-associativity

![image](https://user-images.githubusercontent.com/4564579/162084976-13b1dc56-c7ff-41fe-a4b0-e78faa43ad09.png)

Parentheses are kept when there is a difference in precedence, and thus the actual parsing might not match what the user thinks is happening.  For example: `a + b * c`.  A simple left-to-right reading of this might be read as `(a + b) * c` instead of the correct `a + (b * c)`.  

For the above, where there is no precedence issue if you have `x / y * z` then the simple left to right reading of `(x / y) * z` matches what the language already does, so there is no difference and no need to clarify.
